libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/pulsar/PulsarChangeSubscriber.java

```diff
@@ -0,0 +1,64 @@
+package com.yahoo.athenz.common.messaging.pulsar;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.yahoo.athenz.common.messaging.ChangeSubscriber;
+import com.yahoo.athenz.common.messaging.pulsar.client.AthenzPulsarClient;
+import org.apache.pulsar.client.api.Consumer;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.SubscriptionType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+
+public class PulsarChangeSubscriber<T> implements ChangeSubscriber<T>, Runnable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  protected Consumer<byte[]> pulsarConsumer;
+  protected java.util.function.Consumer<T> processor;
+  protected Class<T> valueType;
+
+  public PulsarChangeSubscriber(String serviceUrl,
+                                String topicName,
+                                String subscriptionName,
+                                SubscriptionType subscriptionType,
+                                AthenzPulsarClient.TlsConfig tlsConfig) {
+    pulsarConsumer = AthenzPulsarClient.createConsumer(serviceUrl,
+        Collections.singleton(topicName),
+        subscriptionName,
+        subscriptionType,
+        tlsConfig
+    );
+    LOG.debug("created publisher: {}, pulsarConsumer: {}", this.getClass(), pulsarConsumer);
+  }
+
+  @Override
+  public void init(java.util.function.Consumer<T> processor, Class<T> valueType) {
+    this.processor = processor;
+    this.valueType = valueType;
+  }
+
+  @Override
+  public void run() {
+    while (true) {
+      LOG.debug("looping over the consumer receive method");
+      try {
+        Message<byte[]> msg = this.pulsarConsumer.receive(1, TimeUnit.SECONDS);
+        if (msg != null) {
+          LOG.debug("received message: {}", new String(msg.getData()));
```

Comment:
surround with "if (LOG.isDebugEnabled())" to avoid creation of string in production (new String(msg.getData()))